pyscf/nac/sacasscf.py

```diff
@@ -0,0 +1,285 @@
+    log.debug ('CASSCF active-orbital gradient:\n{}'.format (de))
+    log.timer ('CASSCF active-orbital gradient', *t0)
+    return de
+
+def gen_g_hop_active (mc, mo, ci0, eris, verbose=None):
+    '''Compute the active-electron part of the orbital rotation gradient
+    by patching out the appropriate block of eris.vhf_c'''
+    moH = mo.conj ().T
+    ncore = mc.ncore
+    vnocore = eris.vhf_c.copy ()
+    vnocore[:,:ncore] = -moH @ mc.get_hcore () @ mo[:,:ncore]
+    with lib.temporary_env (eris, vhf_c=vnocore):
+        return newton_casscf.gen_g_hop (mc, mo, ci0, eris, verbose=verbose)
+
+def _nac_csf (mol, mf_grad, tm1, atmlst):
+    if atmlst is None: atmlst = list (range (mol.natm))
+    aoslices = mol.aoslice_by_atom ()
+    s1 = mf_grad.get_ovlp (mol)
+    # if libcint documentation is to be trusted, mf_grad.get_ovlp
+    # corresponds to differentiating on the SECOND index: <p|dq/dR>
+    nac = np.zeros ((len(atmlst), 3))
+    for k, ia in enumerate (atmlst):
+        shl0, shl1, p0, p1 = aoslices[ia]
+        nac[k] += 0.5*np.einsum ('xij,ij->x', s1[:,p0:p1], tm1[p0:p1])
+    return nac
+
+def nac_csf (mc_grad, mo_coeff=None, ci=None, state=None, mf_grad=None,
+             atmlst=None):
+    '''Compute the "CSF contribution" to the SA-CASSCF NAC'''
+    mc = mc_grad.base
+    if mo_coeff is None: mo_coeff = mc.mo_coeff
+    if ci is None: ci = mc.ci
+    if state is None: state = mc_grad.state
+    if mf_grad is None: mf_grad = mc._scf.nuc_grad_method ()
+    if atmlst is None: atmlst = mc_grad.atmlst
+    mol = mc.mol
+    ket, bra = _unpack_state (state)
+    ncore, ncas, nelecas = mc.ncore, mc.ncas, mc.nelecas
+    castm1 = direct_spin1.trans_rdm1 (ci[bra], ci[ket], ncas, nelecas)
+    # if PySCF commentary is to be trusted, trans_rdm1[p,q] is
+    # <bra|q'p|ket>. I want <bra|p'q - q'p|ket>.
+    castm1 = castm1.conj ().T - castm1
+    mo_cas = mo_coeff[:,ncore:][:,:ncas]
+    tm1 = reduce (np.dot, (mo_cas, castm1, mo_cas.conj ().T))
+    return _nac_csf (mol, mf_grad, tm1, atmlst)
+
+class NonAdiabaticCouplings (sacasscf_grad.Gradients):
+    '''SA-CASSCF non-adiabatic couplings (NACs) between states
+
+    kwargs/attributes:
+
+    state : tuple of length 2
+        The NACs returned are <state[1]|d(state[0])/dR>.
+        In other words, state = (ket, bra).
+    mult_ediff : logical
+        If True, returns NACs multiplied by the energy difference.
+        Useful near conical intersections to avoid numerical problems.
+    use_etfs : logical
+        If True, use the ``electron translation factors'' of Fatehi and
+        Subotnik [JPCL 3, 2039 (2012)], which guarantee conservation of
+        total electron + nuclear momentum when the nuclei are moving
+        (i.e., in non-adiabatic molecular dynamics). This corresponds
+        to omitting the so-called ``CSF contribution'' [cf. JCTC 12,
+        3636 (2016)].
+    '''
+
+    def __init__(self, mc, state=None, mult_ediff=False, use_etfs=False):
```

Comment:
`assert len(state) == 2`

Reply:
if `state` is just an integer, the code is currently set-up to evaluate 

$$
  \langle\psi_I|\nabla_{R}\psi_I\rangle 
$$

which, since the derivative couplings are anti-symmetric with respect to index swapping

$$
\langle\psi_I|\nabla_{R}\psi_J\rangle = - \langle\psi_J|\nabla_{R}\psi_I\rangle
$$

means that we should just return 0. Currently it will return `+/-inf` since division by 0.

> There may need a check to ensure the specified states within the SACASSCF states

Currently, an `IndexError` will occur in `self.make_fcasscf_nacs(state)`. 

I think based off of this, adding the `assert len(state) == 2` as well as a check to make sure that `state[0]` and `state[1]` are some valid, distinct roots should address these problems.

Reply:
Okay, added the following: 

- `assert len(state) == 2` in `__init__ and `kernel` (as well as whenever `_unpack_state` is called.
- when `state[0]==state[1]`, return the correctly shaped 0 object without doing any actual calculations.